# 4.5 Data Consistency Checks
## Most Common Checks
    1) Mixed Data Types, 2) Missing Values, 3) Duplicate Records
## create a dataframe
## create a mixed column type
### Exported datasets 'df_ords_checked.csv' & 'df_prods_checked.csv'

# Import libraries

In [36]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [37]:
path = r'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [38]:
path

'/Users/nancykray/Desktop/Instacart Basket Analysis'

In [39]:
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

In [40]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

##  4.5 Lesson scripts

In [41]:
#create a dataframe
df_test = pd.DataFrame()

In [42]:
# create a mixed column type
df_test['mix'] = ['a','b',1, True]

In [43]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [44]:
# check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_66370/4090404572.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


#### Note: upon later use of the code above, I had to take out 'apply' and leave as just '.map' - may run into this in the future (watch the indents above.  they have to line up right)
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [45]:
# find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [46]:
#create a subset of the dataframe containing only these 16 product_names with missing values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [47]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [48]:
#find number of rows/columns prior to creating a dataframe subset (since these are str values -no mean or median)
df_prods.shape

(49693, 5)

In [49]:
#create new dataframe 'df_prods_clean' for false (finding non-missing values)
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [50]:
#compare number of rows/columns with that which had the missing values (above)
df_prods_clean.shape

(49677, 5)

### this contains exactly 16 less rows than original dataframe (perfect)

In [51]:
#create subset 'df_dups' to find full duplicate rows in 'df_prods_clean' dataframe 
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [52]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [53]:
#before addressing duplicates, check number of rows/columns in datasets and compare
df_prods_clean.shape

(49677, 5)

In [54]:
#create dataframe that doesn't include the duplicates using 'drop_duplicates' function
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [55]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

In [56]:
#2Run the df.describe() function on your df_ords dataframe & look for anything to be investigated further (min/max)
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

### the min is 0 for columns: "Unnamed", "orders_day_of_week", "order_time", & "days_since_prior_order", implying there are missing values as it is impossible to have a 0 day of week or time.
### the max for "orders_days_of_week" is 6, but there are 7 days in a week.  Does this imply that 0. is the index, representing the first day of the week?
### "days_since_prior_order" has max of 30 (monthly?)
### "Order_number" has a min of 1 and max of 100 ( a big difference, especially with a mean of 10)
### the results for "order_id" and "user_id' are not useful as they should be string data type

In [57]:
#3 check for mixed data types in df_ords data frame
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_66370/1330781404.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_66370/1330781404.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_66370/1330781404.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/fb/2jrdrrb52mv7jbhxt0nbx_740000gn/T/ipykernel_66370/1330781404.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

In [58]:
weird

0          False
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

### these columns are False, meaning only one data type exists (no mixed data types)
### #4 can be skipped as only one data type exists and doesn't need to be fixed


In [59]:
#5 run a check for missing values in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

### there are 206209 missing values in "days_since_prior_order" column.  This is a large quantity of numbers and should not be removed.  There is likely an explanation for these missing values (maybe it was the first order that was entered, and therefore has no days prior to the order.  This information should be preserved, as it is valuable.

In [60]:
#6 address the missing values using an appropriate method and explain why
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [61]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_time  days_since_prior_order  
0                 8                     NaN  
11               11                     NaN  
26               14                     NaN  
39               11                     NaN  
45               12                     NaN  
...             ...                     ...  
3420930          12                     NaN  
3420934          18                     NaN  
3421002          19                     NaN  
3421019          15                     NaN  
3421069          11                     NaN  

[206209 rows x 7 columns]

### created a subset dataframe called 'df_ords_nan' to show a table with the NaN or missing values, since there is plenty of other valuable information we need for these orders (and this is far too large to impute a mean or median without greatly impacting the data)

In [62]:
#find number of rows
df_ords.shape

(3421083, 7)

In [63]:
#create new dataframe 'df_ords_clean' for false (finding non-missing values)
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull()==False]

In [64]:
df_ords_clean.shape

(3214874, 7)

### By calculating 342103-3214874=206209, we have verified that there are 206209 less rows in the 'df_ords_clean' compared to 'df_ords' dataframe (amount of missing values has now been verified correctly)

In [65]:
#checking the mean of our column in questions 'days_since_prior_order'
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

### mean is 10

In [66]:
#checking the median
df_ords.median()

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
order_time                     13.0
days_since_prior_order          7.0
dtype: float64

### Median is 7

In [67]:
#checking data types 
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_time                  int64
days_since_prior_order    float64
dtype: object

In [74]:
#flagging column of Nan (using CF Forum to learn how to create a flag column)
df_ords['first_order']=np.where(np.isnan(df_ords['days_since_prior_order'].values), 'First Order', 'Repeat Customer')

In [76]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_time  days_since_prior_order      first_order  
0           8                     NaN      First Order  
1           7                    15.0  Repeat Customer  
2          12                    21.0  Repeat Customer  
3           7                    29.0  Repeat Customer  
4          15                    28.0  Repeat Customer

### #6 notes: I ultimately chose to flag the column and not impute anything since it didn't seem to make sense to use mean or median in its place as it is likely missing only because it there are no days since prior order.  Imputing numbers would change the data.  I did go through the steps above to have practice and see if anything jumped out, but ultimately did not use that information.

In [68]:
#7 Run a check for duplicate values in your df_ords data, and 
# in a markdown, report your findings and propose an explanation for any duplicate values you find.

df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [69]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order]
Index: []

In [81]:
#making sure there were no duplicates in the original data frame too
df_dups2 = df_ords[df_ords.duplicated()]

In [82]:
df_dups2

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order, first_order]
Index: []

### no duplicates found for df_ords

# Export cleaned dataframes with new names

In [83]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords_checked.csv'))

In [84]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','df_prods_checked.csv'))